File: app/src/main/java/com/github/willjgriff/ethereumwallet/data/disk/sharedprefs/SharedPreferencesManager.java
package com.github.willjgriff.ethereumwallet.data.disk.sharedprefs;

import android.content.SharedPreferences;

import com.google.gson.Gson;
import com.google.gson.reflect.TypeToken;

/**
 * Created by Will on 03/04/2016.
 */
public class SharedPreferencesManager {

    private SharedPreferences mSharedPreferences;
    private Gson mGson;

    public SharedPreferencesManager(SharedPreferences sharedPreferences, Gson gson) {
        mSharedPreferences = sharedPreferences;
        mGson = gson;
    }

    public <T> void writeObjectToPreferences(String key, T value) {
        String json = mGson.toJson(value);
        mSharedPreferences.edit().putString(key, json).apply();
    }

    public <T> T readObjectFromPreferences(String key, Class<T> returnType) {
        return mGson.fromJson(mSharedPreferences.getString(key, ""), returnType);
    }

    public <T> T readComplexObjectFromPreferences(String key, TypeToken<T> typeToken) {
        return mGson.fromJson(mSharedPreferences.getString(key, ""), typeToken.getType());
    }

    public void removeObjectFromPreferences(String key) {
        mSharedPreferences.edit().remove(key).apply();
    }
}

File: app/src/main/kotlin/com/github/willjgriff/ethereumwallet/data/SharedPreferencesManager.kt
package com.github.willjgriff.ethereumwallet.data

import android.content.SharedPreferences
import com.google.gson.Gson
import com.google.gson.reflect.TypeToken

/**
 * Created by williamgriffiths on 24/04/2017.
 */
class SharedPreferencesManager(private val sharedPreferences: SharedPreferences, private val gson: Gson) {

    fun <T> writeObjectToPreferences(key: String, value: T) {
        synchronized(this, {
            val json = gson.toJson(value)
            sharedPreferences.edit().putString(key, json).apply()
        })
    }

    fun <T> readObjectFromPreferences(key: String, returnType: Class<T>): T? {
        return synchronized(this, {
            val readObject = sharedPreferences.getString(key, "")
            gson.fromJson(readObject, returnType)
        })
    }

    fun <T> readComplexObjectFromPreferences(key: String, typeToken: TypeToken<T>): T? {
        return synchronized(this, {
            val readObject = sharedPreferences.getString(key, "")
            gson.fromJson<T>(readObject, typeToken.type)
        })
    }

    fun removeObjectFromPrefs(key: String) {
        synchronized(this, { sharedPreferences.edit().remove(key).apply() })
    }
}